File: tuneapi/apis/model_gemini.py
"""
Connect to the Google Gemini API to their LLMs. See more `Gemini <https://ai.google.dev>`_.
"""

# Copyright © 2024-2025 Frello Technology Private Limited
# Copyright © 2025-2025 Yash Bonde github.com/yashbonde
# MIT License
# https://ai.google.dev/gemini-api/docs/function-calling

import httpx
from pydantic import BaseModel
from typing import get_args, get_origin, List, Optional, Dict, Any, Union, Tuple

import tuneapi.utils as tu
import tuneapi.types as tt
from tuneapi.apis.turbo import distributed_chat, distributed_chat_async


class Gemini(tt.ModelInterface):

    def __init__(
        self,
        id: Optional[str] = "gemini-2.0-flash-exp",
        base_url: str = "https://generativelanguage.googleapis.com/v1beta/models/{id}:{rpc}",
        extra_headers: Optional[Dict[str, str]] = None,
        api_token: Optional[str] = None,
        emebdding_url: Optional[str] = None,
    ):
        super().__init__()

        self.model_id = id
        self.base_url = base_url
        self.api_token = api_token or tu.ENV.GEMINI_TOKEN("")
        self.extra_headers = extra_headers
        self.embedding_url = emebdding_url or base_url
        self.client = None

    def set_api_token(self, token: str) -> None:
        self.api_token = token

    def _process_header(self):
        return {
            "Content-Type": "application/json",
        }

    def _process_input(
        self,
        chats: tt.Thread | str,
        model: Optional[str] = None,
        max_tokens: int = 4096,
        temperature: float = 1,
        token: Optional[str] = None,
        debug: bool = False,
        extra_headers: Optional[Dict[str, str]] = None,
        **kwargs,
    ):
        if not token and not self.api_token:  # type: ignore
            raise Exception(
                "Gemini API key not found. Please set GEMINI_TOKEN environment variable or pass through function"
            )
        if isinstance(chats, tt.Thread):
            thread = chats
        elif isinstance(chats, str):
            thread = tt.Thread(tt.human(chats))
        else:
            raise Exception("Invalid input")

        system = ""
        if thread.chats[0].role == tt.Message.SYSTEM:
            system = thread.chats[0].value

        messages = []
        prev_fn_name = ""
        for m in thread.chats[int(system != "") :]:
            if m.role == tt.Message.HUMAN:
                inline_objects = [{"text": m.value}]
                for img in m.images:
                    inline_objects.append(
                        {
                            "inline_data": {
                                "mime_type": "image/png",
                                "data": img,
                            },
                        }
                    )
                messages.append({"role": "user", "parts": inline_objects})
            elif m.role == tt.Message.GPT:
                inline_objects = [{"text": m.value}]
                for img in m.images:
                    inline_objects.append(
                        {
                            "inline_data": {
                                "mime_type": "image/png",
                                "data": img,
                            },
                        }
                    )
                messages.append({"role": "model", "parts": inline_objects})
            elif m.role == tt.Message.FUNCTION_CALL:
                _m = tu.from_json(m.value) if isinstance(m.value, str) else m.value
                prev_fn_name = _m["name"]
                messages.append(
                    {
                        "role": "model",
                        "parts": [
                            {
                                "functionCall": {
                                    "name": _m["name"],
                                    "args": _m["arguments"],
                                }
                            }
                        ],
                    }
                )
            elif m.role == tt.Message.FUNCTION_RESP:
                # _m = tu.from_json(m.value) if isinstance(m.value, str) else m.value
                messages.append(
                    {
                        "role": "function",
                        "parts": [
                            {
                                "functionResponse": {
                                    "name": prev_fn_name,
                                    "response": {
                                        "name": prev_fn_name,
                                        "content": m.value,
                                    },
                                }
                            }
                        ],
                    }
                )
            else:
                raise Exception(f"Unknown role: {m.role}")

        # create headers and params
        headers = self._process_header()
        url = self.base_url.format(
            id=model or self.model_id,
            rpc="streamGenerateContent",
        )

        # create the body
        tools = []
        if isinstance(chats, tt.Thread) and chats.tools:
            tools = [x.to_dict() for x in chats.tools]
        extra_headers = extra_headers or self.extra_headers
        if extra_headers:
            headers.update(extra_headers)

        data = {
            "systemInstruction": {
                "parts": [{"text": system}],
            },
            "contents": messages,
            "safetySettings": [
                {
                    "category": "HARM_CATEGORY_HARASSMENT",
                    "threshold": "BLOCK_MEDIUM_AND_ABOVE",
                },
                {
                    "category": "HARM_CATEGORY_HATE_SPEECH",
                    "threshold": "BLOCK_MEDIUM_AND_ABOVE",
                },
                {
                    "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
                    "threshold": "BLOCK_MEDIUM_AND_ABOVE",
                },
                {
                    "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
                    "threshold": "BLOCK_MEDIUM_AND_ABOVE",
                },
            ],
        }

        generation_config = {
            "temperature": temperature,
            "maxOutputTokens": max_tokens,
            "stopSequences": [],
        }

        if isinstance(chats, tt.Thread) and chats.schema:
            generation_config.update(
                {
                    "response_mime_type": "application/json",
                    "response_schema": get_structured_schema(chats.schema),
                }
            )
        data["generationConfig"] = generation_config

        if tools:
            data["tool_config"] = {
                "function_calling_config": {
                    "mode": "ANY",
                }
            }
            std_tools = []
            for i, t in enumerate(tools):
                props = t["parameters"]["properties"]
                t_copy = t.copy()
                if not props:
                    t_copy.pop("parameters")
                std_tools.append(t_copy)
            data["tools"] = [{"function_declarations": std_tools}]
        data.update(kwargs)

        if debug:
            fp = "sample_gemini.json"
            tu.logger.info("Saving gemini prompt at " + fp)
            tu.to_json(data, fp=fp)

        return url, headers, data

    def _process_output(self, raw: bool, lines_fn: callable):
        block_lines = ""
        done = False
        for line in lines_fn():
            if isinstance(line, bytes):
                line = line.decode("utf-8")

            # get the clean line for block
            if line == ("[{"):  # first line
                line = line[1:]
            elif line == "," or line == "]":  # intermediate or last line
                continue
            block_lines += line

            done = False
            try:
                tu.from_json(block_lines)
                done = True
            except Exception as e:
                pass

            # print(f"{block_lines=}")
            if done:
                part_data = tu.from_json(block_lines)["candidates"][0]["content"][
                    "parts"
                ][0]
                if "text" in part_data:
                    if raw:
                        yield b"data: " + tu.to_json(
                            {
                                "object": "gemini_text",
                                "choices": [{"delta": {"content": part_data["text"]}}],
                            },
                            tight=True,
                        ).encode()
                        yield b""
                    else:
                        yield part_data["text"]
                elif "functionCall" in part_data:
                    fn_call = part_data["functionCall"]
                    fn_call["arguments"] = fn_call.pop("args")
                    yield fn_call
                block_lines = ""

    # Interaction methods

    def stream_chat(
        self,
        chats: tt.Thread | str,
        model: Optional[str] = None,
        max_tokens: int = None,
        temperature: float = 1,
        token: Optional[str] = None,
        timeout=(5, 60),
        usage: bool = False,
        extra_headers: Optional[Dict[str, str]] = None,
        debug: bool = False,
        raw: bool = False,
        **kwargs,
    ):
        url, headers, data = self._process_input(
            chats=chats,
            model=model,
            max_tokens=max_tokens,
            temperature=temperature,
            token=token,
            debug=debug,
            extra_headers=extra_headers,
            **kwargs,
        )

        if self.client is None:
            self.set_client()

        response = self.client.post(
            url=url,
            headers=headers,
            params={"key": self.api_token},
            json=data,
            timeout=timeout,
        )
        try:
            response.raise_for_status()
        except Exception as e:
            yield response.text
            raise e

        yield from self._process_output(
            raw=raw,
            lines_fn=response.iter_lines,
        )

    def chat(
        self,
        chats: tt.Thread | str,
        model: Optional[str] = None,
        max_tokens: int = None,
        temperature: float = 1,
        token: Optional[str] = None,
        usage: bool = False,
        extra_headers: Optional[Dict[str, str]] = None,
        debug: bool = False,
        timeout=(5, 60),
        **kwargs,
    ) -> Any:
        output = ""
        x = None
        try:
            for x in self.stream_chat(
                chats=chats,
                model=model,
                max_tokens=max_tokens,
                temperature=temperature,
                token=token,
                timeout=timeout,
                extra_headers=extra_headers,
                debug=debug,
                raw=False,
                **kwargs,
            ):
                if isinstance(x, dict):
                    output = x
                else:
                    output += x
        except httpx.HTTPError as e:
            print(e.response.text)
            raise e

        if isinstance(chats, tt.Thread) and chats.schema:
            output = chats.schema(**tu.from_json(output))
            return output
        return output

    async def stream_chat_async(
        self,
        chats: tt.Thread | str,
        model: Optional[str] = None,
        max_tokens: int = 4096,
        temperature: float = 1,
        token: Optional[str] = None,
        raw: bool = False,
        debug: bool = False,
        extra_headers: Optional[Dict[str, str]] = None,
        timeout=(5, 60),
        **kwargs,
    ):
        url, headers, data = self._process_input(
            chats=chats,
            model=model,
            max_tokens=max_tokens,
            temperature=temperature,
            token=token,
            debug=debug,
            extra_headers=extra_headers,
            **kwargs,
        )

        if self.async_client is None:
            self.set_async_client()

        response = await self.async_client.post(
            url=url,
            headers=headers,
            params={"key": self.api_token},
            json=data,
            timeout=timeout,
        )
        try:
            response.raise_for_status()
        except Exception as e:
            yield str(e)
            return

        async for chunk in response.aiter_bytes():
            for x in self._process_output(
                raw=raw,
                lines_fn=chunk.decode("utf-8").splitlines,
            ):
                yield x

    async def chat_async(
        self,
        chats: tt.Thread | str,
        model: Optional[str] = None,
        max_tokens: int = None,
        temperature: float = 1,
        token: Optional[str] = None,
        extra_headers: Optional[Dict[str, str]] = None,
        debug: bool = False,
        timeout=(5, 60),
        **kwargs,
    ) -> Any:
        output = ""
        x = None
        try:
            async for x in self.stream_chat_async(
                chats=chats,
                model=model,
                max_tokens=max_tokens,
                temperature=temperature,
                token=token,
                timeout=timeout,
                extra_headers=extra_headers,
                raw=False,
                debug=debug,
                **kwargs,
            ):
                if isinstance(x, dict):
                    output = x
                else:
                    output += x
        except Exception as e:
            if not x:
                raise e
            else:
                raise ValueError(x)

        if isinstance(chats, tt.Thread) and chats.schema:
            output = chats.schema(**tu.from_json(output))
            return output
        return output

    def distributed_chat(
        self,
        prompts: List[tt.Thread],
        post_logic: Optional[callable] = None,
        max_threads: int = 10,
        retry: int = 3,
        pbar: bool = True,
        debug: bool = False,
        time_metrics: bool = False,
        **kwargs,
    ):
        return distributed_chat(
            self,
            prompts=prompts,
            post_logic=post_logic,
            max_threads=max_threads,
            retry=retry,
            pbar=pbar,
            debug=debug,
            time_metrics=time_metrics,
            **kwargs,
        )

    async def distributed_chat_async(
        self,
        prompts: List[tt.Thread],
        post_logic: Optional[callable] = None,
        max_threads: int = 10,
        retry: int = 3,
        pbar: bool = True,
        debug: bool = False,
        time_metrics: bool = False,
        **kwargs,
    ):
        return await distributed_chat_async(
            self,
            prompts=prompts,
            post_logic=post_logic,
            max_threads=max_threads,
            retry=retry,
            pbar=pbar,
            debug=debug,
            time_metrics=time_metrics,
            **kwargs,
        )

    # Embedding methods

    def _prepare_embedding_input(
        self,
        chats: tt.Thread | List[str] | str,
        model: str = "text-embedding-004:embedContent",
        extra_headers: Optional[Dict[str, str]] = None,
    ):
        headers = self._process_header()
        extra_headers = extra_headers or self.extra_headers
        if extra_headers:
            headers.update(extra_headers)
        if isinstance(chats, tt.Thread):
            text = []
            for i, m in enumerate(chats.chats):
                x = f"<{m.role}> : {m.value}"
                text.append(x)
        elif isinstance(chats, tt.Message):
            text = [chats.value]
        elif isinstance(chats, list) and len(chats) and isinstance(chats[0], str):
            if len(chats) > 1:
                raise ValueError(
                    f"Only one string can be passed for list of strings. Got: {len(chats)}"
                )
            text = chats
        elif isinstance(chats, str):
            text = [chats]
        else:
            raise ValueError(f"Invalid input type. Got {type(chats)}")
        data = {
            "model": f"models/{model}",
            "content": {"parts": [{"text": x} for x in text]},
        }
        url = self.base_url.format(id=model, rpc="embedContent")
        return url, headers, data

    def embedding(
        self,
        chats: tt.Thread | List[str] | str,
        model: str = "text-embedding-004",
        extra_headers: Optional[Dict[str, str]] = None,
        token: Optional[str] = None,
        timeout: Tuple[int, int] = (5, 60),
        raw: bool = False,
    ) -> tt.EmbeddingGen:
        url, headers, data = self._prepare_embedding_input(
            chats=chats,
            model=model,
            extra_headers=extra_headers,
        )

        if self.client is None:
            self.set_client()

        response = self.client.post(
            url=url,
            headers=headers,
            json=data,
            timeout=timeout,
            params={"key": token or self.api_token},
        )
        try:
            response.raise_for_status()
        except httpx.HTTPError as e:
            tu.logger.error(f"Cannot get emeddings: {response.text}")
            raise e
        except Exception as e:
            tu.logger.error(e)
            raise e
        resp = response.json()
        if raw:
            return resp

        return tt.EmbeddingGen(embedding=[resp["embedding"]["values"]])

    async def embedding_async(
        self,
        chats: tt.Thread | List[str] | str,
        model: str = "text-embedding-004",
        extra_headers: Optional[Dict[str, str]] = None,
        token: Optional[str] = None,
        timeout: Tuple[float, float] = (5.0, 60.0),  # httpx uses float for timeouts
        raw: bool = False,
    ) -> tt.EmbeddingGen:
        url, headers, data = self._prepare_embedding_input(
            chats=chats,
            model=model,
            extra_headers=extra_headers,
        )

        if self.async_client is None:
            self.set_async_client()

        response = await self.async_client.post(
            url=url,
            headers=headers,
            json=data,
            timeout=timeout,
            params={"key": token or self.api_token},
        )
        try:
            response.raise_for_status()  # Raise an exception for bad status codes
        except httpx.HTTPError as e:
            tu.logger.error(f"Cannot get embeddings: {e.response.text}")
            raise
        except Exception as e:
            tu.logger.error(e)
            raise

        resp = response.json()
        if raw:
            return resp

        return tt.EmbeddingGen(embedding=[resp["embedding"]["values"]])


# helpers


def get_structured_schema(model: type[BaseModel]) -> Dict[str, Any]:
    """
    Converts a Pydantic BaseModel to a JSON schema compatible with Gemini API,
    including `anyOf` for optional or union types and handling nested structures correctly.

    Args:
        model: The Pydantic BaseModel class to convert.

    Returns:
        A dictionary representing the JSON schema.
    """

    def _process_field(
        field_name: str, field_type: Any, field_description: str = None
    ) -> dict:
        """Helper function to process a single field."""
        schema = {}
        origin = get_origin(field_type)
        args = get_args(field_type)

        if origin is list:
            schema["type"] = "array"
            if args:
                item_schema = _process_field_type(args[0])
                schema["items"] = item_schema
                if "type" not in item_schema and "anyOf" not in item_schema:
                    schema["items"]["type"] = "object"  # default item type for list
            else:
                schema["items"] = {}
        elif origin is Optional:
            if args:
                inner_schema = _process_field_type(args[0])
                schema["anyOf"] = [inner_schema, {"type": "null"}]
            else:
                schema = {"type": "null"}
        elif origin is dict:
            schema["type"] = "object"
            if len(args) == 2:
                schema["additionalProperties"] = _process_field_type(args[1])
        else:
            schema = _process_field_type(field_type)

        if field_description:
            schema["description"] = field_description
        return schema

    def _process_field_type(field_type: Any) -> dict:
        """Helper function to process the type of a field."""

        origin = get_origin(field_type)
        args = get_args(field_type)

        if field_type is str:
            return {"type": "string"}
        elif field_type is int:
            return {"type": "integer"}
        elif field_type is float:
            return {"type": "number"}
        elif field_type is bool:
            return {"type": "boolean"}
        elif isinstance(field_type, type) and issubclass(field_type, BaseModel):
            return get_structured_schema(field_type)  # Recursive call for nested models
        elif origin is list:
            schema = {"type": "array"}
            if args:
                item_schema = _process_field_type(args[0])
                schema["items"] = item_schema
                if "type" not in item_schema and "anyOf" not in item_schema:
                    schema["items"]["type"] = "object"
            return schema
        elif origin is Optional:
            return _process_field_type(args[0])
        elif origin is dict:
            schema = {"type": "object"}
            if len(args) == 2:
                schema["additionalProperties"] = _process_field_type(args[1])
            return schema
        elif origin is Union:
            return _process_field_type(args[0])
        else:
            return {"type": "string"}  # default any object to string

    schema = {"type": "object", "properties": {}, "required": []}

    for field_name, field in model.model_fields.items():
        field_description = field.description
        if field.is_required():
            schema["required"].append(field_name)

        schema["properties"][field_name] = _process_field(
            field_name, field.annotation, field_description
        )

    if model.__doc__:
        schema["description"] = model.__doc__.strip()
    return schema

File: ts/apis/model_gemini.ts
/**
 * Google Gemini API implementation for TuneAPI
 */

import { GoogleGenerativeAI } from "@google/generative-ai";
import type {
  Thread,
  Message,
  ModelInterface,
  ChatOptions,
  StreamChatOptions,
  DistributedChatOptions,
  Tool,
  ToolCall,
} from "../types/chats";
import {
  MESSAGE_ROLES,
  createThread,
  human,
  functionResp,
  ToolResponse,
} from "../types/chats";
import { ENV } from "../utils/env";
import { schemaToGeminiFormat, parseWithSchema } from "../types/schema";

/**
 * Google Gemini model implementation using official Google Generative AI SDK
 */
export class GeminiModel implements ModelInterface {
  model_id: string;
  api_token: string;
  private client: GoogleGenerativeAI;

  constructor(id: string = "gemini-2.5-flash", api_token?: string) {
    this.model_id = id;
    this.api_token = api_token || ENV.GEMINI_TOKEN("");

    if (!this.api_token) {
      throw new Error(
        "Gemini API key not found. Please set GEMINI_API_KEY or GEMINI_TOKEN environment variable or pass through constructor"
      );
    }

    this.client = new GoogleGenerativeAI(this.api_token);
  }

  /**
   * Set API token
   */
  setApiToken(token: string): void {
    this.api_token = token;
    this.client = new GoogleGenerativeAI(token);
  }

  /**
   * Translate Thread format to Gemini messages format
   * Returns [systemInstruction, messages]
   */
  private translateThread(thread: Thread): [string, any[]] {
    let systemInstruction = "";
    const messages: any[] = [];

    // Check if first message is system
    if (
      thread.chats.length > 0 &&
      thread.chats[0].role === MESSAGE_ROLES.SYSTEM
    ) {
      systemInstruction = thread.chats[0].value as string;
    }

    // Process remaining messages
    const startIdx = systemInstruction ? 1 : 0;
    for (let i = startIdx; i < thread.chats.length; i++) {
      const m = thread.chats[i];

      if (m.role === MESSAGE_ROLES.HUMAN) {
        const parts: any[] = [{ text: m.value as string }];

        // Add images if present
        for (const img of m.images) {
          parts.push({
            inlineData: {
              mimeType: "image/png",
              data: img,
            },
          });
        }

        messages.push({ role: "user", parts });
      } else if (m.role === MESSAGE_ROLES.GPT) {
        const parts: any[] = [{ text: m.value as string }];

        // Add images if present
        for (const img of m.images) {
          parts.push({
            inlineData: {
              mimeType: "image/png",
              data: img,
            },
          });
        }

        messages.push({ role: "model", parts });
      } else if (m.role === MESSAGE_ROLES.FUNCTION_RESP) {
        // Function responses - add model message with functionCall and user message with functionResponse
        const toolResp = m.value as ToolResponse;

        // Add model message with functionCall using raw_body
        messages.push({
          role: "model",
          parts: [toolResp.tool_call.raw_body],
        });

        // Gemini requires response to be an object (Struct), not a primitive
        // Wrap primitives in an object
        let responseContent = toolResp.result;
        if (typeof responseContent !== "object" || responseContent === null) {
          responseContent = { result: responseContent };
        }

        // Add user message with functionResponse
        messages.push({
          role: "user",
          parts: [
            {
              functionResponse: {
                name: toolResp.tool_call.name,
                response: responseContent,
              },
            },
          ],
        });
      } else if (m.role === MESSAGE_ROLES.THINKING) {
        // Thinking content - add as model message
        if (typeof m.value !== "string") {
          throw new Error(
            `THINKING message value must be a string. Got: '${typeof m.value}'`
          );
        }

        messages.push({
          role: "model",
          parts: [{ text: `[THINKING]\n${m.value}` }],
        });
      } else {
        throw new Error(`Unknown role: ${m.role}`);
      }
    }

    return [systemInstruction, messages];
  }

  /**
   * Convert Tool to Gemini function declaration format
   */
  private convertToolsToGeminiFormat(tools: Tool[]): any[] {
    return tools.map((tool) => {
      const functionDecl: any = {
        name: tool.name,
        description: tool.description,
      };

      // Only add parameters if there are any
      if (tool.parameters.length > 0) {
        functionDecl.parameters = {
          type: "object",
          properties: tool.parameters.reduce((acc, param) => {
            acc[param.name] = {
              type: param.type,
              description: param.description,
              ...(param.enumValues && { enum: param.enumValues }),
              ...(param.items && { items: param.items }),
            };
            return acc;
          }, {} as Record<string, any>),
          required: tool.parameters
            .filter((p) => p.required)
            .map((p) => p.name),
        };
      }

      return functionDecl;
    });
  }

  /**
   * Build request configuration for Gemini API
   */
  private buildRequestConfig(
    messages: any[],
    thread: Thread,
    options?: ChatOptions | StreamChatOptions
  ): any {
    const generationConfig: any = {};
    if (options?.max_tokens) {
      generationConfig.maxOutputTokens = options.max_tokens;
    }
    if (options?.temperature !== undefined) {
      generationConfig.temperature = options.temperature;
    }

    // Add schema if present (structured output)
    if (thread.schema) {
      generationConfig.responseMimeType = "application/json";
      generationConfig.responseSchema = schemaToGeminiFormat(thread.schema);
    }

    if (options?.thinking?.include_thoughts) {
      generationConfig.thinkingConfig = {
        includeThoughts: options.thinking.include_thoughts,
      };
      if (options.thinking.budget_tokens) {
        generationConfig.thinkingConfig.thinkingBudget =
          options.thinking.budget_tokens || 8192; // default to 8192 if not provided
      } else {
        generationConfig.thinkingConfig.thinkingBudget = -1;
      }
    } else if (options?.thinking) {
      generationConfig.thinkingConfig = { thinkingBudget: 0 };
    }

    const requestConfig: any = {
      contents: messages,
      generationConfig,
    };

    // Add tools if present
    if (thread.tools && thread.tools.length > 0) {
      const functionDeclarations = this.convertToolsToGeminiFormat(
        thread.tools
      );
      requestConfig.tools = [{ functionDeclarations }];
      requestConfig.toolConfig = {
        functionCallingConfig: {
          mode: "AUTO",
        },
      };
    }

    return requestConfig;
  }

  /**
   * Stream chat with Gemini - yields tool calls as JSON without executing them
   * The caller is responsible for executing tools and updating the thread
   *
   * Note: Gemini thinking models (e.g., gemini-2.0-flash-thinking-exp) automatically
   * include thinking/reasoning process in the response stream. The thoughts appear
   * naturally in the output before the final answer.
   */
  async *streamChat(
    chats: Thread | string,
    options?: StreamChatOptions
  ): AsyncGenerator<string, void, unknown> {
    const thread =
      typeof chats === "string" ? createThread(human(chats)) : chats;
    const [systemInstruction, messages] = this.translateThread(thread);

    const model = this.client.getGenerativeModel({
      model: options?.model || this.model_id,
      systemInstruction: systemInstruction || undefined,
    });

    const requestConfig = this.buildRequestConfig(messages, thread, options);

    try {
      const result = await model.generateContentStream(requestConfig);
      const collectedFunctionCalls: any[] = [];
      let hasToolCalls = false;

      for await (const chunk of result.stream) {
        // Check for thinking content in candidates
        if (chunk.candidates && chunk.candidates[0]) {
          const candidate = chunk.candidates[0];
          if (candidate.content && candidate.content.parts) {
            for (const part of candidate.content.parts) {
              // Check for thought part (Gemini thinking models)
              const partAny = part as any;
              if (partAny.thought) {
                yield `__THINKING__${part.text}`;
              } else if (part.text) {
                yield part.text;
              }
            }
          }
        }

        // Collect function calls
        const functionCalls = chunk.functionCalls();
        if (functionCalls && functionCalls.length > 0) {
          hasToolCalls = true;
          collectedFunctionCalls.push(...functionCalls);
        }
      }

      // If tool calls were made, yield them as JSON with raw_body
      if (hasToolCalls && collectedFunctionCalls.length > 0) {
        const toolCalls = collectedFunctionCalls.map((fc) => ({
          id: fc.name, // Gemini doesn't use IDs, use function name as ID
          name: fc.name,
          arguments: fc.args,
          raw_body: {
            functionCall: {
              name: fc.name,
              args: fc.args,
            },
          },
        }));

        // Yield tool calls as JSON with a special prefix
        yield `__TOOL_CALLS__${JSON.stringify(toolCalls)}`;
      }
    } catch (error: any) {
      throw new Error(`Gemini API Error: ${error.message}`);
    }
  }

  /**
   * Chat with Gemini (non-streaming) - returns tool calls without executing them
   * The caller is responsible for executing tools and updating the thread
   */
  async chat(
    chats: Thread | string,
    options?: ChatOptions
  ): Promise<string | ToolCall[]> {
    const thread =
      typeof chats === "string" ? createThread(human(chats)) : chats;
    const [systemInstruction, messages] = this.translateThread(thread);

    const model = this.client.getGenerativeModel({
      model: options?.model || this.model_id,
      systemInstruction: systemInstruction || undefined,
    });

    const requestConfig = this.buildRequestConfig(messages, thread, options);

    try {
      const result = await model.generateContent(requestConfig);
      const response = result.response;

      // Check if there are function calls - return them without executing, store raw_body
      const functionCalls = response.functionCalls();
      if (functionCalls && functionCalls.length > 0) {
        const toolCalls: ToolCall[] = functionCalls.map((fc) => ({
          id: fc.name, // Gemini doesn't use IDs, use function name as ID
          name: fc.name,
          arguments: fc.args,
          raw_body: {
            functionCall: {
              name: fc.name,
              args: fc.args,
            },
          },
        }));

        return toolCalls;
      }

      // Extract thinking content if present
      let thinkingText = "";
      if (response.candidates && response.candidates[0]) {
        const candidate = response.candidates[0];
        if (candidate.content && candidate.content.parts) {
          for (const part of candidate.content.parts) {
            const partAny = part as any;
            if (partAny.thought) {
              thinkingText += part.text;
            }
          }
        }
      }

      // No tool calls, return the content
      const content = response.text() || "";

      // Combine thinking and content if both present
      const fullContent = thinkingText
        ? `__THINKING__${thinkingText}\n\n${content}`
        : content;

      // Parse with schema if present (use content without thinking)
      if (thread.schema && content) {
        try {
          return parseWithSchema(thread.schema, content);
        } catch (error: any) {
          throw new Error(`Schema validation failed: ${error.message}`);
        }
      }

      return fullContent;
    } catch (error: any) {
      throw new Error(`Gemini API Error: ${error.message}`);
    }
  }

  /**
   * Distributed chat (placeholder for future implementation)
   */
  async distributedChat(
    prompts: Thread[],
    options?: DistributedChatOptions
  ): Promise<any[]> {
    throw new Error("Distributed chat not yet implemented");
  }
}
